grpclb/src/main/java/io/grpc/grpclb/GrpclbNameResolver.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2015 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.grpclb;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import io.grpc.Attributes;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.NameResolver;
+import io.grpc.Status;
+import io.grpc.internal.DnsNameResolver;
+import io.grpc.internal.SharedResourceHolder.Resource;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * A DNS-based {@link NameResolver} with gRPC LB specific add-ons for resolving balancer
+ * addresses via service records.
+ *
+ * @see SecretGrpclbNameResolverProvider
+ */
+final class GrpclbNameResolver extends DnsNameResolver {
+
+  private static final Logger logger = Logger.getLogger(GrpclbNameResolver.class.getName());
+
+  // From https://github.com/grpc/proposal/blob/master/A5-grpclb-in-dns.md
+  private static final String GRPCLB_NAME_PREFIX = "_grpclb._tcp.";
+
+  GrpclbNameResolver(
+      @Nullable String nsAuthority,
+      String name,
+      Args args,
+      Resource<Executor> executorResource,
+      Stopwatch stopwatch,
+      boolean isAndroid) {
+    super(nsAuthority, name, args, executorResource, stopwatch, isAndroid);
+  }
+
+  @Override
+  protected boolean doResolve(Listener2 listener) {
```

Comment:
I'm not wild about how coupled the GrpclbNameResolver is to DnsNameResolver. It looks quite brittle. This method seems to be the biggest issue.

We may want to discuss in person, but it seems we could try to do a delegation strategy or have the caller of this method notify the `listener`. Both of those seem to allow a `super.doResolve()` approach to let the base class do its work (instead of copying it here and treating the base class as a utility class). As it stands, it would be more straight-forward to have a shared utility class instead of `extend`ing.

Reply:
Reworked this method extension. Introduced an struct for grouping components of resolution result, caller of this method will populate items out and invoke the callback.